### Import libraries


In [2]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split, KFold, cross_val_score
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

### Load data

In [139]:
geo0= pd.read_csv('/datasets/geo_data_0.csv')
geo0.info()
geo0.shape

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


(100000, 5)

In [140]:
geo0

id        f0        f1        f2     product
0      txEyH  0.705745 -0.497823  1.221170  105.280062
1      2acmU  1.334711 -0.340164  4.365080   73.037750
2      409Wp  1.022732  0.151990  1.419926   85.265647
3      iJLyR -0.032172  0.139033  2.978566  168.620776
4      Xdl7t  1.988431  0.155413  4.751769  154.036647
...      ...       ...       ...       ...         ...
99995  DLsed  0.971957  0.370953  6.075346  110.744026
99996  QKivN  1.392429 -0.382606  1.273912  122.346843
99997  3rnvd  1.029585  0.018787 -1.348308   64.375443
99998  7kl59  0.998163 -0.528582  1.583869   74.040764
99999  1CWhH  1.764754 -0.266417  5.722849  149.633246

[100000 rows x 5 columns]

In [141]:
## Identify and fill in missing values

## Step 1: check for outliers
geo0.describe()

## Looking at describe method there are no outliers. There are no values greater than max or lower than min.
## There are no missing values 

f0             f1             f2        product
count  100000.000000  100000.000000  100000.000000  100000.000000
mean        0.500419       0.250143       2.502647      92.500000
std         0.871832       0.504433       3.248248      44.288691
min        -1.408605      -0.848218     -12.088328       0.000000
25%        -0.072580      -0.200881       0.287748      56.497507
50%         0.502360       0.250252       2.515969      91.849972
75%         1.073581       0.700646       4.715088     128.564089
max         2.362331       1.343769      16.003790     185.364347

In [142]:
geo1= pd.read_csv('/datasets/geo_data_1.csv')
geo1.info()
geo1.shape


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


(100000, 5)

In [143]:
## Identify and fill in missing values

## Step 1: check for outliers
geo1.describe()

## Looking at describe method there are no outliers. There are no values greater than max or lower than min.
## There are no missing values 

f0             f1             f2        product
count  100000.000000  100000.000000  100000.000000  100000.000000
mean        1.141296      -4.796579       2.494541      68.825000
std         8.965932       5.119872       1.703572      45.944423
min       -31.609576     -26.358598      -0.018144       0.000000
25%        -6.298551      -8.267985       1.000021      26.953261
50%         1.153055      -4.813172       2.011479      57.085625
75%         8.621015      -1.332816       3.999904     107.813044
max        29.421755      18.734063       5.019721     137.945408

In [144]:
geo2= pd.read_csv('/datasets/geo_data_2.csv')
geo2.info()
geo2.shape


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


(100000, 5)

In [145]:
## Identify and fill in missing values

## Step 1: check for outliers
geo2.describe()

## Looking at describe method there are no outliers. There are no values greater than max or lower than min.
## There are no missing values 

f0             f1             f2        product
count  100000.000000  100000.000000  100000.000000  100000.000000
mean        0.002023      -0.002081       2.495128      95.000000
std         1.732045       1.730417       3.473445      44.749921
min        -8.760004      -7.084020     -11.970335       0.000000
25%        -1.162288      -1.174820       0.130359      59.450441
50%         0.009424      -0.009482       2.484236      94.925613
75%         1.158535       1.163678       4.858794     130.595027
max         7.238262       7.844801      16.739402     190.029838

## Region 0

### Train and Test model

In [148]:
# Separate features and target
features = geo0[['f0', 'f1', 'f2']]
target = geo0['product']

# Preprocessing pipeline for numerical features
numerical_transformer = StandardScaler()

# Combine preprocessing into a pipeline
preprocessor = ColumnTransformer(
    transformers=[
        ('num', numerical_transformer, ['f0', 'f1', 'f2'])
    ]
)

# Model
model = LinearRegression()

# Split data into training and validation sets (75%:25%)
X_train, X_val, y_train, y_val = train_test_split(
    features, target, test_size=0.25, random_state=42)

# Combine preprocessing and model in pipeline
model_pipeline = Pipeline(steps=[
    ('preprocessor', preprocessor),
    ('model', model)
])

# Train the model
model_pipeline.fit(X_train, y_train)

# Make predictions on the validation set
predictions_geo0 = model_pipeline.predict(X_val)

# Save predictions and actual values
validation_results = pd.DataFrame({
    'Predictions': predictions_geo0,
    'Actual': y_val
})

# Calculate RMSE
rmse = np.sqrt(mean_squared_error(y_val, predictions_geo0))
print(f"Validation RMSE: {rmse}")

# Calculate average predicted product
avg_predicted_product_geo0 = np.mean(predictions_geo0)
print(f"Avg Predicted Product: {avg_predicted_product_geo0}")

# Cross-validation
cv = KFold(n_splits=5, shuffle=True, random_state=42)
cv_scores = cross_val_score(model_pipeline, features, target, 
                            cv=cv, scoring='neg_root_mean_squared_error')
print(f"Cross-Validation RMSE: {-np.mean(cv_scores)}")

# Analyze results
print("Validation Results Summary:")
print(validation_results.describe())


Validation RMSE: 37.75660035026169
Avg Predicted Product: 92.39879990657768
Cross-Validation RMSE: 37.6934503621939
Validation Results Summary:
        Predictions        Actual
count  25000.000000  25000.000000
mean      92.398800     92.325956
std       23.217563     44.277512
min       -9.836371      0.021781
25%       76.788865     56.300099
50%       92.414871     90.785176
75%      108.036487    128.117571
max      176.536104    185.355615


The model performs well on average, with predicted values (92.40), which matches the actual value of 92.33.  The RMSE (37.7) reflects moderate prediction errors. This suggests that the model's RMSE differ from the actual values by 37.76 units. 


#### profit calcualtions

In [151]:
# Convert predictions to series for profit function
predictions_series = pd.Series(predictions_geo0, index=y_val.index)

# Parameters 

* total_budget = 100_000_000  # Total budget in dollars
* num_wells = 200             # Number of wells
* revenue_per_barrel = 4.50   # Revenue per barrel in dollars

In [153]:
# Define the profit calculation function
def get_profit(targets, predictions_series):
    # Sort predictions and select top 200 wells
    predictions_sorted = predictions_series.sort_values(ascending=False).head(200)
    selected_wells = targets.loc[predictions_sorted.index]
    
    # Calculate the profit
    profit = (4500 * selected_wells.sum()) - 100_000_000
    
    return round(profit, 2)


In [154]:
# Calculate profit for geo0
profit_geo0 = get_profit(y_val, predictions_series)
print(f"Profit for geo0: ${profit_geo0}")

Profit for geo0: $33591411.14


By selecting the top 200 wells based on model predictions, the profit was calcuated as 33,591,411.14 dollars. This indicates that the model identifies high-production wells, resulting in profit after deducting the fixed cost of 100 million dollars. 

#### Bootstrapping

Calculate the risk of loss and the 95% confidence interval of profits.

Parameters:
- targets: Actual reserves (series).
- predictions: Predicted reserves (series).

Returns:
- 95% CI and risk of loss percentage.

In [158]:
# Risk evaluation function
def calc_risk(targets, predictions_series):

    profits = []

    # Combine predictions and targets into a single DataFrame
    combined_df = pd.DataFrame({'predictions': predictions_geo0, 'targets': targets.reset_index(drop=True)})
    
    # Simulate profits 1000 times with bootstrapping
    for i in range(1000):
        target_subsample = combined_df.sample(n=500, replace=True).reset_index(drop=True)
        profits.append(get_profit(target_subsample['targets'], target_subsample['predictions']))
    
    # Convert profits to Pandas Series for statistical analysis
    profits = pd.Series(profits)
    profit_mean = profits.mean()
    
    # Calculate 95% confidence interval
    upper = profits.quantile(0.975)
    lower = profits.quantile(0.025)
    
    # Calculate the risk of loss
    risk = np.mean(profits < 0) * 100
    
    print(f"95% CI interval: {lower}, {upper}")
    print(f"Risk of loss: {risk}%")



In [159]:
calc_risk(y_val, predictions_series)

95% CI interval: -1365543.82675, 9255727.96675
Risk of loss: 7.3%


The bootstrapping results suggest that the profit from randomly selected subsamples (500 wells) will likely fall within this range 95% of the time. The large range indicates variability in the potential profit.

There is a 6.1% chance of incurring a loss in profit when selecting wells based on the model's predictions. 

## Region 1

In [162]:

# Separate features and target
features = geo1[['f0', 'f1', 'f2']]
target = geo1['product']

# Preprocessing pipeline for numerical features
numerical_transformer = StandardScaler()

# Combine preprocessing into a pipeline
preprocessor = ColumnTransformer(
    transformers=[
        ('num', numerical_transformer, ['f0', 'f1', 'f2'])
    ]
)

# Model
model = LinearRegression()

# Results storage
results = []

# Split data into training and validation sets (75%:25%)
X_train, X_val, y_train, y_val = train_test_split(
    features, target, test_size=0.25, random_state=42)

# Combine preprocessing and model in pipeline
model_pipeline = Pipeline(steps=[
    ('preprocessor', preprocessor),
    ('model', model)
])

# Training model
model_pipeline.fit(X_train, y_train)

# Make predictions on the validation set
predictions_geo1 = model_pipeline.predict(X_val)

# Save predictions and actual values
validation_results = pd.DataFrame({
    'Predictions': predictions_geo1,
    'Actual': y_val
})
results.append(validation_results)

# Calculate RMSE
rmse = np.sqrt(mean_squared_error(y_val, predictions_geo1))
print(f"Validation RMSE: {rmse}")

# Calculate average predicted product
avg_predicted_product_geo1 = np.mean(predictions_geo1)
print(f"Avg Predicted Product: {avg_predicted_product_geo1}")

# Cross-validation
cv = KFold(n_splits=5, shuffle=True, random_state=42)
cv_scores = cross_val_score(model_pipeline, features, target, 
                            cv=cv, scoring='neg_root_mean_squared_error')
print(f"Cross-Validation RMSE: {-np.mean(cv_scores)}")

# Analyze results
print("Validation Results Summary:")
print(validation_results.describe())

Validation RMSE: 0.8902801001028839
Avg Predicted Product: 68.7128780391376
Cross-Validation RMSE: 0.8904111873203165
Validation Results Summary:
        Predictions        Actual
count  25000.000000  25000.000000
mean      68.712878     68.725381
std       45.946672     45.945586
min       -2.070621      0.000000
25%       28.578501     30.132364
50%       57.918041     57.085625
75%      109.313612    107.813044
max      139.983277    137.945408


The model performs well on average, with predicted values (68.71), which matches the actual value of 68.73.  
An RMSE of 0.8903 is very small, indicating good predictive performance

#### profit calcualtions

In [165]:
# Convert predictions to series for profit function
predictions_series = pd.Series(predictions_geo1, index=y_val.index)

In [166]:
# Define the profit calculation function
def get_profit(targets, predictions_series):
    # Sort predictions and select top 200 wells
    predictions_sorted = predictions_series.sort_values(ascending=False).head(200)
    selected_wells = targets.loc[predictions_sorted.index]
    
    # Calculate the profit
    profit = (4500 * selected_wells.sum()) - 100_000_000
    
    return round(profit, 2)

In [167]:
# Calculate profit for geo0
profit_geo1 = get_profit(y_val, predictions_series)
print(f"Profit for geo1: ${profit_geo1}")

Profit for geo1: $24150866.97


By selecting the top 200 wells based on model predictions, the profit was calcuated as $24150866.97. This indicates that the model identifies high-production wells, resulting in profit after deducting the fixed cost of $100 million. 

#### Bootstrapping

In [170]:
# Risk evaluation function
def calc_risk(targets, predictions_series):

    profits = []

    # Combine predictions and targets into a single DataFrame
    combined_df = pd.DataFrame({'predictions': predictions_geo1, 'targets': targets.reset_index(drop=True)})
    
    # Simulate profits 1000 times with bootstrapping
    for i in range(1000):
        target_subsample = combined_df.sample(n=500, replace=True).reset_index(drop=True)
        profits.append(get_profit(target_subsample['targets'], target_subsample['predictions']))
    
    # Convert profits to Pandas Series for statistical analysis
    profits = pd.Series(profits)
    profit_mean = profits.mean()
    
    # Calculate 95% confidence interval
    upper = profits.quantile(0.975)
    lower = profits.quantile(0.025)
    
    # Calculate the risk of loss
    risk = np.mean(profits < 0) * 100
    
    print(f"95% CI interval: {lower}, {upper}")
    print(f"Risk of loss: {risk}%")


In [171]:
calc_risk(y_val, predictions_series)

95% CI interval: 329123.9162500001, 8610145.35525
Risk of loss: 1.6%


The bootstrapping results suggest that the profit from randomly selected subsamples (500 wells) will likely fall within this range 95% of the time. The large range indicates variability in the potential profit.

There is a 2.5% chance of incurring a loss in profit when selecting wells based on the model's predictions. 

## Region 2

In [174]:
### Region 2:geo2

# Separate features and target
features = geo2[['f0', 'f1', 'f2']]
target = geo2['product']

# Preprocessing pipeline for numerical features
numerical_transformer = StandardScaler()

# Combine preprocessing into a pipeline
preprocessor = ColumnTransformer(
    transformers=[
        ('num', numerical_transformer, ['f0', 'f1', 'f2'])
    ]
)

# Model
model = LinearRegression()

# Results storage
results = []

# Split data into training and validation sets (75%:25%)
X_train, X_val, y_train, y_val = train_test_split(
    features, target, test_size=0.25, random_state=42)

# Combine preprocessing and model in pipeline
model_pipeline = Pipeline(steps=[
    ('preprocessor', preprocessor),
    ('model', model)
])

# Training model
model_pipeline.fit(X_train, y_train)

# Make predictions on the validation set
predictions_geo2 = model_pipeline.predict(X_val)

# Save predictions and actual values
validation_results = pd.DataFrame({
    'Predictions': predictions_geo2,
    'Actual': y_val
})
results.append(validation_results)

# Calculate RMSE
rmse = np.sqrt(mean_squared_error(y_val, predictions_geo2))
print(f"Validation RMSE: {rmse}")

# Calculate average predicted product
avg_predicted_product_geo2 = np.mean(predictions_geo2)
print(f"Avg Predicted Product: {avg_predicted_product_geo2}")

# Cross-validation
cv = KFold(n_splits=5, shuffle=True, random_state=42)
cv_scores = cross_val_score(model_pipeline, features, target, 
                            cv=cv, scoring='neg_root_mean_squared_error')
print(f"Cross-Validation RMSE: {-np.mean(cv_scores)}")

# Analyze results
print("Validation Results Summary:")
print(validation_results.describe())

Validation RMSE: 40.145872311342174
Avg Predicted Product: 94.77102387765939
Cross-Validation RMSE: 40.05670367540113
Validation Results Summary:
        Predictions        Actual
count  25000.000000  25000.000000
mean      94.771024     95.150999
std       19.927375     44.783220
min       16.196849      0.019327
25%       81.157678     59.666842
50%       94.612618     94.936982
75%      108.357649    130.566313
max      170.529209    190.011722


The model's Validation RMSE of 40.15 and Cross-Validation RMSE of 40.06, which is indicative of consistent performance. The predicted mean (94.77) closely matches the actual mean (95.15), suggetsing good calibration overall. However, the predicted standard deviation (19.93) is much lower than the actual (44.78), suggesting the model underestimates variability. 

#### profit calcualtions

In [177]:
# Convert predictions to series for profit function
predictions_series = pd.Series(predictions_geo2, index=y_val.index)

In [178]:
# Define the profit calculation function
def get_profit(targets, predictions_series):
    # Sort predictions and select top 200 wells
    predictions_sorted = predictions_series.sort_values(ascending=False).head(200)
    selected_wells = targets.loc[predictions_sorted.index]
    
    # Calculate the profit
    profit = (4500 * selected_wells.sum()) - 100_000_000
    
    return round(profit, 2)

In [179]:
# Calculate profit for geo0
profit_geo2 = get_profit(y_val, predictions_series)
print(f"Profit for geo1: ${profit_geo2}")

Profit for geo1: $25985717.59


By selecting the top 200 wells based on model predictions, the profit was calcuated as $25985717.59. This indicates that the model identifies high-production wells, resulting in profit after deducting the fixed cost of $100 million. 

#### Bootstrapping

In [182]:
# Risk evaluation function
def calc_risk(targets, predictions_series):

    profits = []

    # Combine predictions and targets into a single DataFrame
    combined_df = pd.DataFrame({'predictions': predictions_geo2, 'targets': targets.reset_index(drop=True)})
    
    # Simulate profits 1000 times with bootstrapping
    for i in range(1000):
        target_subsample = combined_df.sample(n=500, replace=True).reset_index(drop=True)
        profits.append(get_profit(target_subsample['targets'], target_subsample['predictions']))
    
    # Convert profits to Pandas Series for statistical analysis
    profits = pd.Series(profits)
    profit_mean = profits.mean()
    
    # Calculate 95% confidence interval
    upper = profits.quantile(0.975)
    lower = profits.quantile(0.025)
    
    # Calculate the risk of loss
    risk = np.mean(profits < 0) * 100
    
    print(f"95% CI interval: {lower}, {upper}")
    print(f"Risk of loss: {risk}%")


In [183]:
calc_risk(y_val, predictions_series)

95% CI interval: -1407473.52375, 8971474.664
Risk of loss: 8.0%


The bootstrapping results suggest that the profit from randomly selected subsamples (500 wells) will likely fall within this range 95% of the time. The large range indicates variability in the potential profit.

There is a 8.8% chance of incurring a loss in profit when selecting wells based on the model's predictions. 

### Summary

Geo0 offers the highest profit among the three regions, making it a strong candidate for development. However, profitability alone does not determine the best region; risk and confidence intervals should also be considered.


Geo1 has the lowest risk of loss (2.5%) and a positive lower bound in the 95% confidence interval, indicating a minimal chance of financial loss. Although geo0 provides the highest profit, its higher risk and the possibility of a negative outcome make geo1 a safer and more stable choice.

In [192]:
BUDGET = 100000000            # Budget in USD
NUM_WELLS = 200                 # Number of wells to develop
REVENUE_PER_UNIT = 4500         # Revenue per thousand barrels

# 3.2 break-even volume per well
break_even_production = BUDGET / (REVENUE_PER_UNIT * NUM_WELLS)
print(f"Break-even production per well: {break_even_production:.2f} thousand barrels")


# average reserves per region
avg_reserves = {
    'geo0': avg_predicted_product_geo0,
    'geo1': avg_predicted_product_geo1,
    'geo2': avg_predicted_product_geo2
}

# 3.3 average reserves versus break-even production
print("\nComparison with Break-even Production:")
for region, avg in avg_reserves.items():
    status = "BELOW" if avg < break_even_production else "ABOVE"
    print(f"{region}: Average reserves = {avg} → {status} break-even point")


print("\nFindings:")
print(f"The break-even production per well is {break_even_production:.2f} thousand barrels.")
for region, avg in avg_reserves.items():
    if avg < break_even_production:
        print(f"{region} has average reserves of {avg}, which is BELOW the break-even threshold.")
    else:
        print(f"{region} has average reserves of {avg}, which is ABOVE the break-even threshold.")


Break-even production per well: 111.11 thousand barrels

Comparison with Break-even Production:
geo0: Average reserves = 92.39879990657768 → BELOW break-even point
geo1: Average reserves = 68.7128780391376 → BELOW break-even point
geo2: Average reserves = 94.77102387765939 → BELOW break-even point

Findings:
The break-even production per well is 111.11 thousand barrels.
geo0 has average reserves of 92.39879990657768, which is BELOW the break-even threshold.
geo1 has average reserves of 68.7128780391376, which is BELOW the break-even threshold.
geo2 has average reserves of 94.77102387765939, which is BELOW the break-even threshold.


All three regions (geo0, geo1, and geo2) have average reserves below the break-even threshold, which suggetss that none of the regions can actually cover development costs without more selective well development strategy. 